tests/Authenticator.hs

```diff
@@ -0,0 +1,434 @@
+      forM_ excludeCredentialDescriptorList $ \descriptor -> case authenticatorLookupCredential authenticator (M.pkcdId descriptor) of
+        Just (_, item) -> do
+          when (rpId == pkcsRpId item) (fail "InvalidStateError")
+        Nothing -> pure ()
+
+      -- 4. If requireResidentKey is true and the authenticator cannot store a
+      -- client-side discoverable public key credential source, return an error
+      -- code equivalent to "ConstraintError" and terminate the operation.
+      -- NOTE: We do not have to do this because the test authenticator supports
+      -- both methods of discoverability
+
+      -- 5. If requireUserVerification is true and the authenticator cannot
+      -- perform user verification, return an error code equivalent to
+      -- "ConstraintError" and terminate the operation.
+      -- NOTE: We do not do this because we fake user verification
+
+      -- 6. Collect an authorization gesture confirming user consent for creating
+      -- a new credential. The prompt for the authorization gesture is shown by
+      -- the authenticator if it has its own output capability, or by the user
+      -- agent otherwise. The prompt SHOULD display rpEntity.id, rpEntity.name,
+      -- userEntity.name and userEntity.displayName, if possible.
+      --   If requireUserVerification is true, the authorization gesture MUST
+      --   include user verification.
+      --   If requireUserPresence is true, the authorization gesture MUST include a
+      --   test of user presence.
+      --   If the user does not consent or if user verification fails, return an
+      --   error code equivalent to "NotAllowedError" and terminate the operation.
+      -- NOTE: We curently always succeed this step
+
+      -- 7. Once the authorization gesture has been completed and user consent
+      -- has been obtained, generate a new credential object:
+      -- 7.1. Let (publicKey, privateKey) be a new pair of cryptographic keys
+      -- using the combination of PublicKeyCredentialType and cryptographic
+      -- parameters represented by the first item in credTypesAndPubKeyAlgs that
+      -- is supported by this authenticator.
+      (publicKey, privateKey) <- newKeyPair $ M.pkcpAlg param
+
+      -- 7.2. Let userHandle be userEntity.id.
+      let userHandle = M.pkcueId userEntity
+
+      -- 7.4 If requireResidentKey is true or the authenticator chooses to
+      -- create a client-side discoverable public key credential source:
+
+      -- 7.4.1 Let credentialId be a new credential id.
+      -- NOTE: We need to have a CredentialId before we can construct the
+      -- credentialSource.
+      -- NOTE: We always choose to construct a clientside discoverable
+      -- credential, as this is allowed (See 7.4).
+      -- TODO: Use deterministic random number generator (Uniform)
+      credentialId <- M.CredentialId <$> Random.getRandomBytes 16
```

Comment:
I guess I should reclarify: Both `MonadRandom` and `Uniform` and co. seem to support deterministic and non-deterministic randomness. We should ideally only use one of those in the top-level functions. Also we should make the tests deterministic, which probably requires only using one of them.

Reply:
I think MonadRandom isn't deterministic in this case as it is the IO instance. But yeah, we should only use one of them.